src/main/java/org/links/objecthash/ObjectHash.java

```diff
@@ -0,0 +1,318 @@
+                                               JSONException {
+    digester.reset();
+    digester.update((byte) ('l'));
+    for (int n = 0; n < list.length(); ++n) {
+      ObjectHash innerObject = new ObjectHash();
+      innerObject.hashAny(list.get(n));
+      digester.update(innerObject.hash());
+    }
+    hash = digester.digest();
+  }
+
+  private String debugString(Iterable<ByteBuffer> buffers) {
+    StringBuilder sb = new StringBuilder();
+    for (ByteBuffer buff : buffers) {
+      sb.append('\n');
+      sb.append(toHex(buff.array()));
+    }
+    return sb.toString();
+  }
+
+  private void hashObject(JSONObject obj) throws NoSuchAlgorithmException, 
+                                                 JSONException {
+    List<ByteBuffer> buffers = new ArrayList<ByteBuffer>();
+    Comparator<ByteBuffer> ordering = new Comparator<ByteBuffer>() {
+        @Override
+        public int compare(ByteBuffer left, ByteBuffer right) {
+          for (int idx = 0; idx < left.array().length; ++idx) {
+            if (left.array()[idx] < right.array()[idx]) {
+              return -1;
+            } else if (left.array()[idx] > right.array()[idx]) {
+              return 1;
+            }
+          }
+          return 0;
+        }          
+    };
+    Iterator<String> keys = obj.keys();
+    while (keys.hasNext()) {
+      ByteBuffer buff = ByteBuffer.allocate(2 * SHA256_BLOCK_SIZE);
+      String key = keys.next();
+      // TODO(phad): would be nice to chain all these calls builder-stylee.
+      ObjectHash hKey = new ObjectHash();
+      hKey.hashString(key);
+      ObjectHash hVal = new ObjectHash();
+      hVal.hashAny(obj.get(key));
+      buff.put(hKey.hash());
+      buff.put(hVal.hash());
+      buffers.add(buff);
+    }
+    LOG.info("Accumulated ByteBuffers: " + debugString(buffers));
+    Collections.sort(buffers, ordering);
+    LOG.info("Sorted ByteBuffers: " + debugString(buffers));
+    digester.reset();
+    digester.update((byte) 'd');
+    for (ByteBuffer buff : buffers) {
+      digester.update(buff.array());
+    }
+    hash = digester.digest();
+  }
+
+  private static int parseHex(char digit) {
+    assert ((digit >= '0' && digit <= '9') || (digit >= 'a' && digit <= 'f'));
+    if (digit >= '0' && digit <= '9') {
+      return digit - '0';
+    } else {
+      return 10 + digit - 'a';
+    }
+  }
+
+  public static ObjectHash fromHex(String hex) throws NoSuchAlgorithmException {
+    ObjectHash h = new ObjectHash();
+    hex = hex.toLowerCase();
+    if (hex.length() % 2 == 1) {
+      hex = '0' + hex;
+    }
+    // TODO(phad): maybe just use Int.valueOf(s).intValue()
+    int pos = SHA256_BLOCK_SIZE;
+    for (int idx = hex.length(); idx > 0; idx -= 2) {
+      h.hash[--pos] = (byte) (16 * parseHex(hex.charAt(idx - 2))
+                              + parseHex(hex.charAt(idx - 1)));
+    }
+    return h;
+  }
+
+  private static JsonType getType(Object jsonObj) {
```

Comment:
I suspect you really want to follow the design pattern I'm slowly introducing in the Python:

https://github.com/benlaurie/objecthash/blob/master/objecthash.py#L180